File: quizzes3/10-problem.ts
export {};

const findFirstRepeatChar = (str: string): string | null => {
    const seen = new Set<string>();

    for (const char of str) {
        if (seen.has(char)) return char;
        seen.add(char);
    }

    return null;
};

const findAllRepeatChar = (str: string): string[] => {
    const counts = new Map<string, number>();
    const duplicates: string[] = [];

    for (const char of str) {
        counts.set(char, (counts.get(char) ?? 0) + 1);
    }

    for (const [char, count] of counts) {
        if (count > 1) duplicates.push(char);
    }

    return duplicates;
};

console.log(JSON.stringify({ data: findFirstRepeatChar('swiss') }, null, 2));
console.log(JSON.stringify({ data: findFirstRepeatChar('them') }, null, 2));
console.log(
    JSON.stringify({ data: findAllRepeatChar('programming') }, null, 2),
);

File: bins8/10-problem.rb
# frozen_string_literal: true

require 'json'

def find_first_repeat_char(str)
  seen = {}

  str.each_char do |char|
    return char if seen[char]
    seen[char] = true
  end

  nil
end

def find_all_repeat_chars(str)
  counts = Hash.new(0)

  str.each_char { |c| counts[c] += 1 }

  counts.select { |_, count| count > 1 }.keys
end

puts JSON.pretty_generate({ data: find_first_repeat_char('swiss') })
puts JSON.pretty_generate({ data: find_first_repeat_char('them') })
puts JSON.pretty_generate({ data: find_all_repeat_chars('programming') })
